[LoopUnroll] Consider convergence control tokens when unrolling

ssahasra opened the pull request:
- There is no restriction on a loop with controlled convergent operations when
  the relevant tokens are defined and used within the loop.

- When a token defined outside a loop is used inside (also called a loop
  convergence heart), unrolling is allowed only in the absence of remainder or
  runtime checks.

- When a token defined inside a loop is used outside, such a loop is said to be
  "extended". This loop can only be unrolled by also duplicating the extended part
  lying outside the loop. Such unrolling is disabled for now.

- Clean up loop hearts: When unrolling a loop with a heart, duplicating the
  heart will introduce multiple static uses of a convergence control token in a
  cycle that does not contain its definition. This violates the static rules for
  tokens, and needs to be cleaned up into a single occurrence of the intrinsic.

- Spell out the initializer for UnrollLoopOptions to improve readability.


Original implementation [D85605] by Nicolai Haehnle <nicolai.haehnle@amd.com>.

arsenm commented on llvm/include/llvm/Analysis/CodeMetrics.h:
```diff
@@ -45,6 +46,9 @@ struct CodeMetrics {
   /// True if this function contains a call to a convergent function.
   bool convergent = false;
 
+  /// True if the code contains an uncontrolled convergent operation.
+  bool convergentUncontrolled = false;
```
Maybe combine these into a ConvergenceType field?

ssahasra replied:
Fixed.

efriedma-quic commented on llvm/lib/Analysis/CodeMetrics.cpp:
```diff
@@ -111,11 +112,22 @@ void CodeMetrics::collectEphemeralValues(
 
+static bool isUsedOutsideOfLoop(const Instruction &I, const Loop &L) {
+  for (const auto *U : I.users()) {
+    if (auto *I = dyn_cast<Instruction>(U)) {
```
All users of Instructions are Instructions; just `cast<>`.

ssahasra replied:
Fixed.

efriedma-quic commented on llvm/lib/Transforms/Utils/LoopUnroll.cpp:
```diff
@@ -731,6 +740,16 @@ llvm::UnrollLoop(Loop *L, UnrollLoopOptions ULO, LoopInfo *LI,
           NewPHI->eraseFromParent();
         }
 
+        // Eliminate copies of the loop heart intrinsic, if any.
```
If I understand the rules here correctly, this is only necessary if you're not completely unrolling the loop?

ssahasra replied:
Correct. If the loop is completely unrolled, then the loop intrinsic is redundant. Some other pass should eventually clean that up. But when a loop is partially unrolled, those copies of the heart violate the static rules, and all of them should be replaced by the instance from the header.

efriedma-quic commented on llvm/lib/Analysis/CodeMetrics.cpp:
```diff
@@ -163,20 +175,34 @@ void CodeMetrics::analyzeBasicBlock(
+        LLVM_DEBUG(dbgs() << I
+                          << "\n  Cannot duplicate a convergence control token "
+                             "used outside the loop.\n");
+        notDuplicatable = true;
```
This seems like it's overloading the meaning of "notDuplicatable"; maybe consider splitting the convergence stuff into a separate bit?

ssahasra replied:
The meaning of "notDuplicatable" is not overloaded; it's exactly what the name says, right? It's just that now a block may be "not duplicatable" for two different reasons. But we don't really have any clients that need to distinguish why something is notDuplicatable. We could revisit this variable when such a distinction becomes necessary?

efriedma-quic replied:
FunctionSpecialization is currently using CodeMetrics; does it really not care about the distinction?

arsenm replied:
I still think we should get rid of noduplicate entirely. It was a failed attempt to avoid inventing convergence

ssahasra replied:
Fixed. `canUnroll()` now separately considers any discovered convergence instead of mixing it with `notDuplicatable`.

efriedma-quic commented:
Latest version looks fine to me, but I'll let @arsenm give final approval.

ssahasra committed:
[LoopUnroll] Consider convergence control tokens when unrolling

- There is no restriction on a loop with controlled convergent operations when
  the relevant tokens are defined and used within the loop.

- When a token defined outside a loop is used inside (also called a loop
  convergence heart), unrolling is allowed only in the absence of remainder or
  runtime checks.

- When a token defined inside a loop is used outside, such a loop is said to be
  "extended". This loop can only be unrolled by also duplicating the extended part
  lying outside the loop. Such unrolling is disabled for now.

- Clean up loop hearts: When unrolling a loop with a heart, duplicating the
  heart will introduce multiple static uses of a convergence control token in a
  cycle that does not contain its definition. This violates the static rules for
  tokens, and needs to be cleaned up into a single occurrence of the intrinsic.

- Spell out the initializer for UnrollLoopOptions to improve readability.

Original implementation [D85605] by Nicolai Haehnle <nicolai.haehnle@amd.com>.
